src/colvarmodule.cpp

```diff
@@ -794,7 +808,9 @@ int colvarmodule::calc()
       // Write restart file, if different from main output
       error_code |= write_restart_file(restart_out_name);
     } else {
-      error_code |= write_restart_file(output_prefix()+".colvars.state");
+      if (output_prefix().size()) {
+        error_code |= write_restart_file(output_prefix()+".colvars.state");
```

Comment:
Suggestion: use a different extension for binary restart files (`colvars.binstate`?), to make it obvious to the user and avoid mistakes.

Reply:
I looked into this for a bit, and the `.colvars.state` extension is used pervasively throughout the code. Making another extension work consistently would require significant restructuring, which I would rather not add to this PR.

My thinking behind not distinguishing the two was the following:
- LAMMPS and GROMACS have their own checkpoint/restart files and do not need `.colvars.state` files in any format;
- NAMD does not have a restart file, however NAMD users are dealing with the `.coor` file being in either PDB or binary format, so being confused would actually be consistent behavior!

Would you be okay with not supporting silent auto-detection of the format but raise instead an error when the file being read is not in the expected format? Whether you accidentally fail, or consciously neglect to set the environment variable in consecutive jobs, a scream test should be the appropriate course of action.

Reply:
Okay, thinking about it some more, I can't think of a case where a mistake would have serious consequences. So I'm going to backtrack on my comment, I think it works as it is.

Reply:
New idea: split the magic number into two halves. One to specify the file type, and one for the version. That way we can print a useful error message when trying to read an incompatible binary file.

Reply:
The main problem is that the code version may not be enough to ensure that the binary file is compatible, as it could have been made assuming a different memory layout for complex types.  Safer to just declare "compatibility not guaranteed" in my opinion.

As far as the version of Colvars that produced the file, that's encoded in the `configuration { ... }` string block, which is the very first thing that appears both in text files and in binary files (after the magic number).